packages/flutter/lib/src/rendering/selection.dart

```diff
@@ -104,17 +113,164 @@ abstract class SelectionHandler implements ValueListenable<SelectionGeometry> {
   SelectionResult dispatchSelectionEvent(SelectionEvent event);
 }
 
+/// This class stores the information of an active selection under
+/// a [Selectable] or [SelectionHandler].
+///
+/// The [SelectedContentRange]s for a given [Selectable] or
+/// [SelectionHandler] can be retrieved by calling
+/// [SelectionHandler.getSelections].
+///
+/// [SelectionArea] and [SelectableRegion] provide access to the
+/// [SelectedContentRange]s that represent their active selection
+/// through the [SelectionListener.onSelectionChanged] callback.
+///
+/// See also:
+///
+///   * [SelectionListener], which provides selection updates in
+///   the form of [SelectedContentRange]s for the subtree it wraps,
+///   contained under a [SelectionArea] or [SelectableRegion].
+@immutable
+class SelectedContentRange {
+  /// Creates a [SelectedContentRange] with the given values.
+  const SelectedContentRange({
+    this.selectableId,
+    required this.contentLength,
+    required this.startOffset,
+    required this.endOffset,
+    this.children,
```

Comment:
why does this have children? shouldn't the SelectedContentRange in SelectionDetails.ranges just a flatten list?

Reply:
I assume this is for widgetspan, but if they care about selection inside the widgetspan, they should provide a selectionId in the widgetspan themselves

Reply:
I think the reason I had this as a tree was because a flat list felt discontinuous to me. By that I mean a single `Text` widget could have multiple `SelectedContentRange`s if there were multiple `WidgetSpan`s that broke it up vs with a tree structure where each `Text` widget would have a single `SelectedContentRange` associated with it and the children of that range are any sub-selections within it.

For example, for the text widget below, with a flattened list selecting all the text would return 3 `SelectedContentRange`s vs with a tree it only returns 1 SelectedContentRange with 1 child range:

```dart
const Text.rich(
 TextSpan(
   text: 'Hello world, ',
   children: <InlineSpan>[
     WidgetSpan(
       child: Text('how are you today?'),
     ),
     TextSpan(
       text: 'More text after widgetspan',
     ),
   ],
 ),
)
```

I'm not sure which approach is better? but I think it makes more sense that a `Text` widget only reports 1 `SelectedContentRange` since the user doesn't really expect a `Text` widgets selection to be split into multiple segments.

Reply:
I don't think we should reflect selection range structure with text span tree unless we want to couple selectionListener with text widget. Otherwise, it would be confusing that a certain datamodel will produce selectedcontentrange format in a certain way. It feels like magic.

 other way is if we find a way to couple datamodel and selection id, maybe we can tag selection id on per textspan basis.

Another thing is whether the widget span should be included. It feels weird and tricky to include that as one selected content range since widget span can be anything.

Reply:
> I don't think we should reflect selection range structure with text span tree unless we want to couple selectionListener with text widget. Otherwise, it would be confusing that a certain datamodel will produce selectedcontentrange format in a certain way. It feels like magic.

I was more thinking this reflected a structure where a leaf widget that is selectable might have children selectables. In most cases this children list will be empty, but in the cases where a selectable does have children then a user could look into this list to see more details. A user could always choose not to look into the children list and only look at the root.

Another issue that gets trickier is ids, if we do return multiple `SelectedContentRange`s for a single selectable widget then a user will have to likely need to map multiple ranges to a single id. Tagging selection id on per textspan basis would solve this but it seems strange to me to do this in cases when there is no widgetspan. For example say a user has a textspan tree, with multiple textspans it seems odd to require them to tag each individual textspan when the underlying selectable tree would only create one selectable for this case.

> Another thing is whether the widget span should be included. It feels weird and tricky to include that as one selected content range since widget span can be anything.

Can you explain more on this? If the widgetspan contains a column of N selectable widgets, then there would be N `SelectedContentRange`s reported.